presto-open-telemetry/src/main/java/com/facebook/presto/opentelemetry/OpenTelemetryTracerProvider.java

```diff
@@ -0,0 +1,119 @@
+import com.facebook.presto.spi.tracing.TracerHandle;
+import com.facebook.presto.spi.tracing.TracerProvider;
+import com.facebook.presto.tracing.TracingConfig;
+import com.google.inject.Inject;
+
+import java.util.Map;
+import java.util.function.Function;
+
+import static com.facebook.presto.opentelemetry.OpenTelemetryHeaders.PRESTO_B3_SINGLE_HEADER_PROPAGATION;
+import static com.facebook.presto.opentelemetry.OpenTelemetryHeaders.PRESTO_BAGGAGE_HEADER;
+import static com.facebook.presto.opentelemetry.OpenTelemetryHeaders.PRESTO_TRACE_TOKEN;
+import static com.facebook.presto.opentelemetry.OpenTelemetryHeaders.PRESTO_W3C_PROPAGATION;
+import static com.facebook.presto.spi.StandardErrorCode.CONTEXT_PROPAGATOR_ERROR;
+
+public class OpenTelemetryTracerProvider
+        implements TracerProvider
+{
+    @Inject
+    public OpenTelemetryTracerProvider() {}
+
+    @Override
+    public String getName()
+    {
+        return "Open telemetry tracer provider";
+    }
+
+    @Override
+    public String getTracerType()
+    {
+        return "otel";
+    }
+
+    @Override
+    public Function<Map<String, String>, TracerHandle> getHandleGenerator()
+    {
+        Function<Map<String, String>, TracerHandle> handleGenerator = new Function<Map<String, String>, TracerHandle>()
+        {
+            @Override
+            public TracerHandle apply(Map<String, String> headers)
+            {
+                String contextPropagator = determineContextPropagationMode(headers);
+                return new OpenTelemetryTracerHandle(
+                        headers.get(PRESTO_TRACE_TOKEN),
+                        contextPropagator,
+                        getPropagatedContextFromHeader(contextPropagator, headers),
+                        headers.get(PRESTO_BAGGAGE_HEADER));
+            }
+        };
+        return handleGenerator;
+    }
+
+    @Override
+    public Tracer getNewTracer(TracerHandle handle)
+    {
+        OpenTelemetryTracerHandle tracerHandle = (OpenTelemetryTracerHandle) handle;
+        return new OpenTelemetryTracer(
+                tracerHandle.getTraceToken(),
+                tracerHandle.getContextPropagator(),
+                tracerHandle.getPropagatedContext(),
+                tracerHandle.getBaggage());
+    }
+
+    /**
+     * Take header values and determine which context propagation mode is used
+     * Currently only supports b3 single header
+     * @param headers HTTP request headers
+     * @return context propagation mode
+     */
+    private String determineContextPropagationMode(Map<String, String> headers)
+    {
+        if (headers.containsKey(PRESTO_B3_SINGLE_HEADER_PROPAGATION)) {
+            return "b3_single_header";
+        }
+        else if (headers.containsKey(PRESTO_W3C_PROPAGATION)) {
+            throw new PrestoException(
+                    CONTEXT_PROPAGATOR_ERROR,
+                    "Only b3 single header context propagation mode is currently supported.");
```

Comment:
two options:
- you can create your own error code class within the module with a brand new base code; check any other implementation of `ErrorCodeSupplier`; or
- you can just throw runtime error or put generic code to be a catch-all to save the effort, which is not recommended but fine